internal/store/in-memory-store.go

```diff
@@ -0,0 +1,276 @@
+			}
+			// Use the last part as the tag
+			tag := tagParts[len(tagParts)-1]
+			// Get the digest for the tag
+			digest, err := s.fetcher.GetDigest(ctx, tag)
+			if err != nil {
+				return nil, err
+			}
+
+			// Check if the image exists and matches the digest
+			if !(s.checkImageAndDigest(digest, img.Name)) {
+				change = true
+			}
+
+		}
+
+		// Create imageMap filled with all images from remote imageList
+		imageMap := make(map[string]bool)
+		for _, img := range imageList {
+			imageMap[img.Name] = true
+		}
+
+		// Iterate over in memory store and remove any image that is not found in imageMap
+		for digest, image := range s.images {
+			if _, exists := imageMap[image]; !exists {
+				s.Remove(ctx, digest, image)
+				change = true
+			}
+		}
+		// Print out the entire store for debugging purposes
+		fmt.Println("Current store:")
+		for digest, imageRef := range s.images {
+			fmt.Printf("Digest: %s, Image: %s\n", digest, imageRef)
+		}
+
+	}
+	if change {
+		fmt.Println("Changes detected in the store")
+		change = false
+		return imageList, nil
+	} else {
+		fmt.Println("No changes detected in the store")
+		return nil, nil
+	}
+
+}
+
+func (s *inMemoryStore) Add(ctx context.Context, digest string, image string) error {
+	// Check if the image already exists in the store
+	if _, exists := s.images[digest]; exists {
+		fmt.Printf("Image: %s, digest: %s already exists in the store.\n", image, digest)
+		return fmt.Errorf("image %s already exists in the store", image)
+	} else {
+		// Add the image and its digest to the store
+		s.images[digest] = image
+		fmt.Printf("Successfully added image: %s, digest: %s\n", image, digest)
+		return nil
+	}
+}
+
+func (s *inMemoryStore) AddImage(ctx context.Context, image string) error {
+	// Add the image to the store
+	s.images[image] = ""
+	fmt.Printf("Added image: %s\n", image)
+	return nil
+}
+
+func (s *inMemoryStore) Remove(ctx context.Context, digest string, image string) error {
+	// Check if the image exists in the store
+	if _, exists := s.images[digest]; exists {
+		// Remove the image and its digest from the store
+		delete(s.images, digest)
+		fmt.Printf("Successfully removed image: %s, digest: %s\n", image, digest)
+		return nil
+	} else {
+		fmt.Printf("Failed to remove image: %s, digest: %s. Not found in the store.\n", image, digest)
+		return fmt.Errorf("image %s not found in the store", image)
+	}
+}
+
+func (s *inMemoryStore) RemoveImage(ctx context.Context, image string) error {
+	// Remove the image from the store
+	delete(s.images, image)
+	fmt.Printf("Removed image: %s\n", image)
+	return nil
+}
+
+// checkImageAndDigest checks if the image exists in the store and if the digest matches the image reference
+func (s *inMemoryStore) checkImageAndDigest(digest string, image string) bool {
+
+	// Check if the received image exists in the store
+	for storeDigest, storeImage := range s.images {
+		if storeImage == image {
```

Comment:
I prefer to inverse the check and call continue instead. This is just to avoid the amount of statement nesting that is done. Improves readability.